### PR TITLE
Allow different HardBreaks settings for documents and comments

### PR DESCRIPTION
**Pull request by zeripath**
GH has different HardBreaks behaviour for markdown comments and documents.

Comments have hard breaks and documents have soft breaks - therefore Gitea's rendering will always be different from GH's if we only provide one setting.

Here we split the setting in to two - one for documents and one for comments and other things.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: web_src/js/index.js [hunk 1]
```diff
@@ -65,14 +65,18 @@ function initEditPreviewTab($form) {
     $previewTab.on('click', function () {
       const $this = $(this);
       let context = `${$this.data('context')}/`;
+      let mode = this.data('markdown-mode');
+      if (!mode) {
+        mode = 'comment';
+      }
```
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
      const mode = this.data('markdown-mode') || 'comment';
```

File: web_src/js/index.js [hunk 2]
```diff
@@ -65,14 +65,15 @@ function initEditPreviewTab($form) {
     $previewTab.on('click', function () {
       const $this = $(this);
       let context = `${$this.data('context')}/`;
+      const mode = this.data('markdown-mode') || 'comment';
```
Comment (silverwind) [inline web_src/js/index.js]: I guess this should actually be `$this.data`? Or just use `this.dataset.markdownMode`.

Comment (silverwind) [reply web_src/js/index.js]: Or maybe you need to select for the `textarea` that has the attribute as `this` should refer the tab here, not the textarea.

Comment (zeripath) [reply web_src/js/index.js]: It genuinely worked in testing! I've put the gfm mode on the preview tab.

Comment (silverwind) [reply web_src/js/index.js]: Looking good now.

**Comment by guillep2k** [comment]
Unit test panics:

image

I don't understand why on Earth copying text in a drone log is so 🤬 difficult.

**Comment by zeripath** [comment]
@guillep2k unit tests fixed

File: custom/conf/app.ini.sample [hunk 3]
```diff
@@ -216,7 +216,11 @@ EVENT_SOURCE_UPDATE_TIME = 10s
 ; Render soft line breaks as hard line breaks, which means a single newline character between
 ; paragraphs will cause a line break and adding trailing whitespace to paragraphs is not
 ; necessary to force a line break.
-ENABLE_HARD_LINE_BREAK = true
+; Previously one setting `ENABLE_HARD_LINE_BREAK` covered both
```
Comment (guillep2k) [inline custom/conf/app.ini.sample]:
```suggestion
; Previously one setting `ENABLE_HARD_LINE_BREAK` covered both cases.
```
Although I'm not sure this line even belongs here.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -152,9 +152,12 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 
 ## Markdown (`markdown`)
 
-- `ENABLE_HARD_LINE_BREAK`: **true**: Render soft line breaks as hard line breaks, which
+- `ENABLE_HARD_LINE_BREAK_IN_COMMENTS`: **true**: Render soft line breaks as hard line breaks in comments, which
   means a single newline character between paragraphs will cause a line break and adding
-  trailing whitespace to paragraphs is not necessary to force a line break.
+  trailing whitespace to paragraphs is not necessary to force a line break. (Previously one setting covered both.)
```
Comment (guillep2k) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
  trailing whitespace to paragraphs is not necessary to force a line break.
```
I think it's weird to explain this two times.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 5]
```diff
@@ -152,9 +152,12 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 
 ## Markdown (`markdown`)
 
-- `ENABLE_HARD_LINE_BREAK`: **true**: Render soft line breaks as hard line breaks, which
+- `ENABLE_HARD_LINE_BREAK_IN_COMMENTS`: **true**: Render soft line breaks as hard line breaks in comments, which
```
Comment (guillep2k) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]:
```suggestion
Note: before 1.13, a single setting (`ENABLE_HARD_LINE_BREAK`) was used to control the following two cases simultaneously. It was deprecated in favor of better granularity.
- `ENABLE_HARD_LINE_BREAK_IN_COMMENTS`: **true**: Render soft line breaks as hard line breaks in comments, which
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 6]
```diff
@@ -152,9 +152,12 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
...
   means a single newline character between paragraphs will cause a line break and adding
-  trailing whitespace to paragraphs is not necessary to force a line break.
+  trailing whitespace to paragraphs is not necessary to force a line break. (Previously one setting covered both.)
+- `ENABLE_HARD_LINE_BREAK_IN_DOCUMENTS`: **false**: Render soft line breaks as hard line breaks in documents, which
+  means a single newline character between paragraphs will cause a line break and adding
+  trailing whitespace to paragraphs is not necessary to force a line break. (Previously one setting covered both.)
```
Comment (guillep2k) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
  trailing whitespace to paragraphs is not necessary to force a line break.
```

File: routers/api/v1/misc/markdown.go [hunk 7]
```diff
@@ -61,7 +63,14 @@ func Markdown(ctx *context.APIContext, form api.MarkdownOption) {
 			}
 		}
 		if ctx.Repo != nil && ctx.Repo.Repository != nil {
-			meta = ctx.Repo.Repository.ComposeMetas()
+			if form.Mode == "gfm" {
```
Comment (guillep2k) [inline routers/api/v1/misc/markdown.go:67]:
```suggestion
			// "gfm" = GitHub Flavored Markdown, as used by the javascript library
			if form.Mode == "gfm" {
```
Replace "the javascript library" with the library name 😁

File: modules/markup/markdown/goldmark.go [hunk 8]
```diff
@@ -151,6 +151,16 @@ func (g *ASTTransformer) Transform(node *ast.Document, reader text.Reader, pc pa
 					v.AppendChild(v, newChild)
 				}
 			}
+		case *ast.Text:
+			if v.SoftLineBreak() && !v.HardLineBreak() {
+				renderMetas := pc.Get(renderMetasKey).(map[string]string)
+				mode := renderMetas["mode"]
+				if mode != "document" {
+					v.SetHardLineBreak(setting.Markdown.EnableHardLineBreakInComments)
```
Comment (guillep2k) [inline modules/markup/markdown/goldmark.go:159]: Should this be set on each node? 🤔

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 5 above)
Comment (zeripath) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]: It's not even deprecated - it's just not used.

File: modules/markup/markdown/goldmark.go (hunk 8 above)
Comment (zeripath) [reply modules/markup/markdown/goldmark.go:159]: yes.

The text nodes are rendered by: 

https://github.com/yuin/goldmark/blob/a302193b064875a8af8cd241985cb26574f37408/renderer/html/html.go#L602-L623

I cannot adjust the renderer as you do not have the parsercontext at that point.

File: routers/api/v1/misc/markdown.go (hunk 7 above)
Comment (zeripath) [reply routers/api/v1/misc/markdown.go:67]: Library? It's what's the API uses...

```suggestion
			// "gfm" = Github Flavored Markdown - set this to render as a document
			if form.Mode == "gfm" {
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 5 above)
Comment (lafriks) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]: As this was introduced in 1.12 we still have time to fix this to not be breaking

Comment (zeripath) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]: oh cool, let's remove that comment then!

File: custom/conf/app.ini.sample [hunk 9]
```diff
@@ -216,7 +216,11 @@ EVENT_SOURCE_UPDATE_TIME = 10s
 ; Render soft line breaks as hard line breaks, which means a single newline character between
 ; paragraphs will cause a line break and adding trailing whitespace to paragraphs is not
 ; necessary to force a line break.
-ENABLE_HARD_LINE_BREAK = true
+; Previously one setting `ENABLE_HARD_LINE_BREAK` covered both cases.
```
Comment (zeripath) [inline custom/conf/app.ini.sample]:
```suggestion
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 5 above)
Comment (lafriks) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]: Default was changed at least, setting was there before but was false by default imho

Comment (silverwind) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:155]: Yes it was https://github.com/go-gitea/gitea/pull/11162

File: routers/api/v1/misc/markdown.go (hunk 7 above)
Comment (guillep2k) [reply routers/api/v1/misc/markdown.go:67]: OK, it was confusing having two different names for the same setting ("gfm" vs "document"). But adding the comment is good enough. 👍

**Review by guillep2k — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
make lg-tm work

**Comment by zeripath** [comment]
@lafriks do I need to add warning to the documentation about the changes?

**Comment by lafriks** [comment]
As we will backport it I don't think it's needed as this will set back previous behavior that we made as breaking change in 1.12.0-rc1

**Comment by zeripath** [comment]
make lg-tm work

**Pull request merged**